File: sdks/typescript/test/sdk-v2-unknown-commit.mjs
// SPDX-License-Identifier: GPL-3.0-only
import assert from "node:assert/strict";

import { HyphaeClient, ProductError } from "../dist/v2/index.js";

const clients = [
  HyphaeClient.local(process.env.HYPHAE_SOCKET),
  HyphaeClient.http(process.env.HYPHAE_ORIGIN, { bearerToken: process.env.HYPHAE_TOKEN }),
];
const errors = [];
for (const [index, client] of clients.entries()) {
  try {
    await client.structureSet(
      new TextEncoder().encode(`unknown-${index}`),
      new TextEncoder().encode("t".repeat(9000)),
      undefined,
      { requestId: 30_132n },
    );
    throw new Error("unknown commit was acknowledged");
  } catch (error) {
    assert(error instanceof ProductError);
    assert.equal(error.fields.code, "unknown_commit");
    assert.equal(error.fields.category, "unavailable");
    assert.equal(error.fields.retry, "unknown-commit");
    assert.equal(error.fields.transactionState, "outcome-unknown");
    assert.notEqual(error.fields.transactionId, undefined);
    assert.equal(error.fields.requestId, 30_132n);
    const { transactionId, details, ...comparable } = error.fields;
    errors.push(comparable);
  }
}
assert.deepEqual(errors[0], errors[1]);
for (const client of clients) await client.close();

File: crates/hyphae-server/tests/fixtures/sdk_v2_unknown_commit.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import os

from hyphae_sdk.v2 import HyphaeClient, ProductError, RequestOptions

clients = [
    HyphaeClient.local(os.environ["HYPHAE_SOCKET"]),
    HyphaeClient.http(os.environ["HYPHAE_ORIGIN"], bearer_token=os.environ["HYPHAE_TOKEN"]),
]
errors = []
for client, row_id in zip(clients, (9, 14)):
    try:
        client.structure_set(
            f"unknown-{row_id}".encode(), b"p" * 9000,
            options=RequestOptions(request_id=20_132),
        )
        raise AssertionError("unknown commit was acknowledged")
    except ProductError as error:
        assert error.code == "unknown_commit"
        assert error.category == "unavailable"
        assert error.retry == "unknown-commit"
        assert error.transaction_state == "outcome-unknown"
        assert error.transaction_id is not None
        assert error.request_id == 20_132
        values = dict(error.fields.__dict__)
        values.pop("transaction_id")
        errors.append(values)
assert errors[0] == errors[1]
for client in clients:
    client.close()
